daemons/gptp/common/ptp_message.cpp

```diff
@@ -1762,7 +1768,12 @@ void PTPMessageSignalling::sendPort(IEEE1588Port * port, PortIdentity * destIden
 	memcpy(buf_ptr + PTP_SIGNALLING_TARGET_PORT_IDENTITY(PTP_SIGNALLING_OFFSET),
 	       &targetPortIdentify, sizeof(targetPortIdentify));
 
-	tlv.toByteString(buf_ptr + PTP_COMMON_HDR_LENGTH + PTP_SIGNALLING_LENGTH);
+	size_t available_size = sizeof(buf_t) - (PTP_COMMON_HDR_LENGTH + PTP_SIGNALLING_LENGTH
+						+ port->getPayloadOffset());
+	int result = tlv.toByteString(buf_ptr + PTP_COMMON_HDR_LENGTH + PTP_SIGNALLING_LENGTH,
+						 available_size);
+	if(result != 0)
+		GPTP_LOG_ERROR("toByteString returned an error - invalid size passed");
 
 	port->sendGeneralPort(PTP_ETHERTYPE, buf_t, messageLength, MCAST_OTHER, destIdentity);
```

Comment:
This got to be rather lengthy.  I want to start out saying that I appreciate the effort and initiative in making the pull request and this only got as long as it did because of underlying stuff I want to have a conversation about, not because I want to be critical of the code in the pull request!  It's a great implementation (with a little bit more needed) of the requested change, but as I looked deeper I started to question some assumptions behind it.

**Regarding the pull request as-is**

Do we still want to send (here and in the other `sendPort` method) if the `tlv.toByteString` calls fail?  The headers at this point include the length of the tlv, but we haven't written any tlv data, and we couldn't write it because there wasn't enough room, so the length in our header exceeds the size of the actual payload buffer. We also pass `messageLength` (which includes the tlv length) to our port's send routine, so it will happily read beyond `buf_t` (which as a side note is a poor name for a variable; POSIX reserves all names ending in `_t`, and the ones it uses are for *types*, not variables, so it's also confusing) into other parts of the stack.

**Thoughts on the general approach**

If I understand correctly, the only real non-static length we're dealing with here is `port->getPayloadOffset()`, and the return value of that method should never change for the lifetime of the port. Ideally we'd make a static assertion that our static buffer size is big enough manage the largest possible getPayloadOffset() value for all the different port types, but I don't think there's a way to express that sort of thing directly in C++.

Second best would be a dynamic assertion along with a unit test that combines calls to this with all the possible port types.  But we're definitely not there with the unit test infrastructure on gPTP yet.

In any case, I'm not sure adding the length checking to `toByteString` is the right thing to do.  Certainly length checking *does* need to be done, but a note in the documentation for `toByteString` that it assumes the caller has already done length checking would be sufficient. It is often useful for performance reasons to have low-level stream writing primitives that don't do checks like that so that you can do one higher-level check followed by a fixed-length series of the lower-level writer calls that don't check. If you check in every primitive, all the checks can add up to a significant portion of packet construction time, and most of them are redundant. It is something of a performance vs. mistake-proofing trade-off, but the code generally seems to follow this approach and it would take a lot of rework to move it to a more 'mistake-proof' approach. As the relevant code is mostly tested and working, I don't see much value in doing that.

In these TLV cases, a single length check per sendPort method should be sufficient to guarantee safety, and I think it's a great thing to add to avoid regressions in the future, though it should probably be done via some sort of assertion or check-and-abort near the beginning of the methods.